### PR TITLE
Simplify module

### PR DESCRIPTION
**Pull request by nbhuiyan**
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

### CODE REVIEW

File: b9/include/b9/Module.hpp [hunk 1]
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
...
-  FunctionDef(const std::string& name, std::uint32_t index,
-              std::vector<Instruction>&& instructions, std::uint32_t nparams = 0,
-              std::uint32_t nlocals = 0)
-      : name{name},
-        index{index},
-        instructions{std::move(instructions)},
-        nparams{nparams},
-        nlocals{nlocals} {}
-
+
```
Comment (rwy7) [inline b9/include/b9/Module.hpp]: The diff looks weird. You added some whitespace, I think. Run clang-format?

File: b9/include/b9/Module.hpp [hunk 2]
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
...
-      : name{name},
-        index{index},
-        instructions{std::move(instructions)},
-        nparams{nparams},
-        nlocals{nlocals} {}
-
+        
   // Function Data
   std::string name;
   uint32_t index;
```
Comment (rwy7) [inline b9/include/b9/Module.hpp]: I don't think the index in FunctionDef is used anywhere--can we just get rid of it?

File: test/testDisasm.cpp [hunk 3]
```diff
@@ -174,14 +174,6 @@ TEST(RoundTripSerializationTest, testFunctionSection) {
...
   roundTripFunctionSection(functions);
 
-  std::vector<Instruction> i2 = {{OpCode::INT_PUSH_CONSTANT, 2},
-                                 {OpCode::INT_PUSH_CONSTANT, 2},
-                                 {OpCode::INT_ADD},
-                                 {OpCode::FUNCTION_RETURN},
-                                 END_SECTION};
-  auto f2 = FunctionDef("testName", 5, i2, 1, 2);
-  functions.push_back(f2);
-  EXPECT_THROW(roundTripFunctionSection(functions), DeserializeException);
```
Comment (rwy7) [inline test/testDisasm.cpp:184]: What was this test even doing? Purposefully getting the index wrong?

File: b9/include/b9/Module.hpp [hunk 4]
```diff
@@ -26,17 +25,7 @@ struct FunctionDef {
         instructions{instructions},
         nparams{nparams},
         nlocals{nlocals} {}
```
Comment (rwy7) [inline b9/include/b9/Module.hpp]: I think we can actually get rid of this constructor too, and just rely on the implicitly defined constructors.

File: b9/include/b9/Module.hpp (hunk 1 above)
Comment (nbhuiyan) [reply b9/include/b9/Module.hpp]: VS code auto indents when adding new line, and does not show up in its diff viewer by default, so I missed it. Clang-format should fix that. Good catch!

File: test/testDisasm.cpp (hunk 3 above)
Comment (nbhuiyan) [reply test/testDisasm.cpp:184]: Yup! The index of the second function is 5, so `readFunctionSection()` in the deserializer throws an exception due to this:
```c++
 for (uint32_t i = 0; i < functionCount; i++) {
     functions.emplace_back("", -1, std::vector<Instruction>{});
     readFunction(in, functions.back());
     if (functions[i].index != i) {
      throw DeserializeException{"Invalid index"};
     }
   }
```

**Comment by nbhuiyan** [comment]
@rwy0717 I have addressed your comments. Also, I ran clang format on a number of files that do not have anything to do with this PR.

File: b9/include/b9/deserialize.hpp [hunk 5]
```diff
@@ -48,9 +48,10 @@ void readStringSection(std::istream &in, std::vector<std::string> &strings);
 
 bool readInstructions(std::istream &in, std::vector<Instruction> &instructions);
 
-void readFunctionData(std::istream &in, FunctionDef &functionSpec);
+void readFunctionData(std::istream &in, FunctionDef &functionSpec,
+                      uint32_t index);
```
Comment (rwy7) [inline b9/include/b9/deserialize.hpp]: I think the index can go from this parameter list.

File: b9/include/b9/deserialize.hpp [hunk 6]
```diff
@@ -48,9 +48,10 @@ void readStringSection(std::istream &in, std::vector<std::string> &strings);
...
 
-void readFunction(std::istream &in, FunctionDef &functionDef);
+void readFunction(std::istream &in, FunctionDef &functionDef, uint32_t index);
```
Comment (rwy7) [inline b9/include/b9/deserialize.hpp]: same here.

File: b9/src/deserialize.cpp [hunk 7]
```diff
@@ -34,18 +34,18 @@ bool readInstructions(std::istream &in,
   return true;
 }
 
-void readFunctionData(std::istream &in, FunctionDef &functionDef) {
+void readFunctionData(std::istream &in, FunctionDef &functionDef,
+                      uint32_t index) {
```
Comment (rwy7) [inline b9/src/deserialize.cpp]: don't need the index.

**Review by rwy7 — CHANGES_REQUESTED** [review]
You added an index parameter to a few APIs, which we now don't need. Otherwise looks good.

**Comment by nbhuiyan** [comment]
Totally forgot about that. That's now taken care of, thanks!

**Review by rwy7 — APPROVED** [review]

**Pull request merged**